pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultipleOrEqualitiesToInClauseFilterQueryTreeOptimizer.java

```diff
@@ -18,163 +18,221 @@
+        children.set(nonEqInTree.getValue(), optimizedTree);
+      }
+      return filterQueryTree;
     }
+  }
 
+  private FilterQueryTree optimizeTreeRootedAtAND(FilterQueryTree filterQueryTree) {
+    applyOptimizationToChildNodes(filterQueryTree);
     return filterQueryTree;
   }
 
   private void applyOptimizationToChildNodes(FilterQueryTree filterQueryTree) {
-    Iterator<FilterQueryTree> childTreeIterator = filterQueryTree.getChildren().iterator();
-    List<FilterQueryTree> childrenToAdd = null;
-
-    while (childTreeIterator.hasNext()) {
-      FilterQueryTree childQueryTree = childTreeIterator.next();
-      FilterQueryTree optimizedChildQueryTree = optimize(childQueryTree, filterQueryTree);
+    List<FilterQueryTree> children = filterQueryTree.getChildren();
+    for (int i = 0; i < children.size(); i++) {
+      FilterQueryTree childQueryTree = children.get(i);
+      FilterQueryTree optimizedChildQueryTree = optimize(childQueryTree);
       if (childQueryTree != optimizedChildQueryTree) {
-        childTreeIterator.remove();
-        if (childrenToAdd == null) {
-          childrenToAdd = new ArrayList<>();
-        }
-        childrenToAdd.add(optimizedChildQueryTree);
+        children.set(i, optimizedChildQueryTree);
       }
     }
-
-    if (childrenToAdd != null) {
-      filterQueryTree.getChildren().addAll(childrenToAdd);
-    }
   }
 
-  private boolean collectChildOperators(FilterQueryTree filterQueryTree, Map<String, Set<String>> columnToValues,
-      List<FilterQueryTree> nonEqualityOperators) {
+  private boolean collectChildOperatorsOfRootOROperator(FilterQueryTree filterQueryTree,
+      Map<String, PredicateInfo> columnToInEqPredicateInfo, Map<FilterQueryTree, Integer> nonEqInOperators) {
     boolean containsDuplicates = false;
-
+    int childIndex = 0;
     for (FilterQueryTree childQueryTree : filterQueryTree.getChildren()) {
-      if (childQueryTree.getOperator() == FilterOperator.EQUALITY
-          || childQueryTree.getOperator() == FilterOperator.IN) {
-        List<String> childValues = valueDoubleTabListToElements(childQueryTree.getValue());
-
-        if (!columnToValues.containsKey(childQueryTree.getColumn())) {
-          TreeSet<String> value = new TreeSet<>(childValues);
-          columnToValues.put(childQueryTree.getColumn(), value);
-          if (!containsDuplicates && value.size() != childValues.size()) {
-            containsDuplicates = true;
-          }
+      FilterOperator operator = childQueryTree.getOperator();
+      if (operator == FilterOperator.EQUALITY || operator == FilterOperator.IN) {
+        List<String> childValues = childQueryTree.getValue();
+        String column = childQueryTree.getColumn();
+        PredicateInfo predicateInfo = columnToInEqPredicateInfo.get(column);
+        if (predicateInfo == null) {
+          predicateInfo = new PredicateInfo(new HashSet<>(), false, operator == FilterOperator.IN);
+          columnToInEqPredicateInfo.put(column, predicateInfo);
```

Comment:
This logic has been simplified and now also collects the position of each non IN/EQUALITY child operator